### PR TITLE
Issue/67/mock_flexibility

### PR DESCRIPTION
**Pull request by combet**
Mock data generation now includes the option of randomly drawing the redshift of background galaxies from a redshift distribution. Demo in examples/fake_data_generation.ipynb has been updated to exemplify the new functionality.

### CODE REVIEW

**Comment by combet** [comment]
Acutally, I just noticed that this lists commits dating back to July 29. But we only did a few commits in the last few days, so I'm not sure what has happened here.

File: clmm/models/cosmo_dependent_models/cosmology_model.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+
+    """
+    print("getting cosmology %s, need to implement options from astropy"%cosmology_str)
+
+    return apycosmo.FlatLambdaCDM(H0=70, Om0=0.3)
+
+
+
+def angular_diameter_distance(redshift, cosmology_str):
```
Comment (m-aguena) [inline clmm/models/cosmo_dependent_models/cosmology_model.py]: Add description of function

File: clmm/models/CLMM_densityModels.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+import colossus.cosmology.cosmology as Cosmology
+import colossus.halo.profile_dk14 as profile_dk14
+import colossus.halo as Halo
+import colossus.halo.concentration as hc
+from clmm.models import Profile1D
+
+
+import numpy as np
+
+
```
Comment (m-aguena) [inline clmm/models/CLMM_densityModels.py]: Please add documentation for all classes and functions defined here

File: clmm/models/dk14.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+# -*- coding: utf-8 -*-
+"""
+Created on Mon Jul 30 22:33:55 2018
+
+@author: MattFong
+"""
+from clmm.models.radial_models.radial_model_1d import RadialModel1D
+import colossus.halo.profile_dk14 as profile_dk14
+
+
```
Comment (m-aguena) [inline clmm/models/dk14.py]: Please add documentation for all classes and functions defined here

File: clmm/models/radial_models/nfw.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+from clmm.models.radial_models.radial_model_1d import RadialModel1D
+
+
+import colossus.cosmology.cosmology as Cosmology
+
+from colossus.halo.mass_so import M_to_R
+from colossus.halo.mass_so import densityThreshold
+
+import numpy as np
+
```
Comment (m-aguena) [inline clmm/models/radial_models/nfw.py]: Please add documentation for all classes and functions defined here

File: tests/test_radial_model_1d.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+
+from __future__ import absolute_import, print_function
+import numpy as np
+from clmm.models.radial_models.radial_model_1d import RadialModel1D
+from colossus.halo.mass_so import M_to_R
+from colossus.halo.mass_so import densityThreshold
+
```
Comment (m-aguena) [inline tests/test_radial_model_1d.py]: Please add documentation for all functions defined here

**Comment by matthewkirby** [comment]
Regarding travis failing. The output is below
```
============================= test session starts ==============================
platform linux -- Python 3.6.3, pytest-3.3.0, py-1.5.2, pluggy-0.6.0
rootdir: /home/travis/build/LSSTDESC/CLMM, inifile:
collected 20 items                                                             
tests/test_clmmbase.py ..                                                [ 10%]
tests/test_galaxycluster.py .......                                      [ 45%]
tests/test_manager.py .......                                            [ 80%]
tests/test_model.py .                                                    [ 85%]
tests/test_radial_model_1d.py F                                          [ 90%]
tests/test_shear_azimuthal_averager.py .                                 [ 95%]
tests/test_travis.py .                                                   [100%]
=================================== FAILURES ===================================
_____________________________ test_radial_model_1d _____________________________
    def test_radial_model_1d():
        r = np.linspace(.05, 3., 15)
    
        sample_profile = RadialModel1D(z_lens=0.25, func=nfwprofile,
                                           params={'M': 1E15, 'c': 4},
>                                          z_source = 1.)
tests/test_radial_model_1d.py:53: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/clmm-0.1-py3.6.egg/clmm/models/radial_models/radial_model_1d.py:66: in __init__
    self.sigma_crit = self.calculate_sigma_crit_with_cosmology()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <clmm.models.radial_models.radial_model_1d.RadialModel1D object at 0x7fc7ae8764a8>
    def calculate_sigma_crit_with_cosmology(self) :
        # calculate da lens
        # calculate da source
        # calculate da lens source
>       d_a_lens = angular_diameter_distance(self.z_lens)
E       TypeError: angular_diameter_distance() missing 1 required positional argument: 'cosmology_str'
/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/clmm-0.1-py3.6.egg/clmm/models/radial_models/radial_model_1d.py:278: TypeError
```

So it is failing in `tests/test_radial_model_1d.py`  and it looks like you are missing an argument for the cosmology object in angular_diameter_distance()

**Comment by combet** [comment]
Thanks @matthewkirby and yes, I forgot to run pytest before the PR :/ 

@matthewwf2001, @mchoi8739, it looks like everything developed in branch Issue/16/interface-models-superclass-with-profiles has been merged with the mock_flexibility branch (hence the failing test and the missing comments @m-aguena is requesting). But as far as I can tell, it is not needed for your addition of the redshift distribution to the mock_data. 

Am I missing something? I'm not sure how to proceed from here...

**Comment by matthewwf2001** [comment]
Sorry about all that! The addition of the probability distribution function is not necessary for mock_data to work--we just added it so we can get an array of redshifts for the sources. The UTD group has been writing papers/dissertation drafts for a deadline so we haven't gotten much change to work on CLMM today. If the code we added is causing problems feel free to delete the section we added and Miyoung and I will add it back in properly at a future time.

**Comment by combet** [comment]
@matthewwf2001 , @mchoi8739  sorry for the confusion here. The bit of code you added for the redshift distribution to mock_data.py is absolutely fine and works great :)  Thanks again for that.

It's just that in clmm/models of this branch, there is extra stuff compared to master (along with a new test), and I believe it comes from the Issue/16/interface-models-superclass-with-profiles branch. I am not sure how it got here but it's not needed. So if that's OK, I will removed all the unnecessary bits so the PR may go through.

**Comment by matthewwf2001** [comment]
The only thing I can think of that we changed recently has been in clmm/models/cosmo_dependent_models/cosmology_model.py, where we fixed the return for angular_diameter_distance_two_objects(...). Outside of that we are okay with anything you want to delete!

**Pull request closed without merging**